source/utils/timed-out.ts

```diff
@@ -31,14 +32,24 @@ export default (request: ClientRequest, delays: Delays, options: TimedOutOptions
 	request[reentry] = true;
 	const cancelers: Array<typeof noop> = [];
 	const {once, unhandleAll} = unhandler();
+	const timers: typeof global =
+		isTestEnv ?
+		// @ts-ignore
+			request.timers || global :
+			global;
+
+	const addTimeout = (delay: number, callback: (...args: unknown[]) => void, type: string, ...args: unknown[]): (typeof noop) => {
+		if (type === 'socket') {
+			request.setTimeout(delay, callback);
+			return noop;
+		}
 
-	const addTimeout = (delay: number, callback: (...args: unknown[]) => void, ...args: unknown[]): (typeof noop) => {
 		// Event loop order is timers, poll, immediates.
 		// The timed event may emit during the current tick poll phase, so
 		// defer calling the handler until the poll phase completes.
 		let immediate: NodeJS.Immediate;
-		const timeout: NodeJS.Timeout = setTimeout(() => {
-			immediate = setImmediate(callback, delay, ...args);
+		const timeout: NodeJS.Timeout = timers.setTimeout(() => {
+			immediate = timers.setImmediate(callback, delay, type, ...args);
```

Comment:
I don't like this approach. You're adding test code into production. These two need to be separated.

Reply:
Use `lolex.install()` in tests (the `before` startup event).

Reply:
`clock.tick()` seems to be proper here.

Reply:
Agreed. Only the test files should change, not the source files.

Reply:
We can add a compile step to remove the test code but the strategy makes sense.

To use `.install()`, you would need to have a single clock instance that is shared across the entire node process. I initially tested with this & it doesn't work because:

  1. It causes issues with ava trying to schedule itself.
  2. All tests must be run serially in order to share a clock instance otherwise you end up with races.

Generally speaking, it is a better practice to use localized clocks since the dependency injection surface area is controlled.

Reply:
I have updated the `request.timers` property to use a symbol instead of an identifier property so that the behaviour can only be triggered from within the package.

Reply:
Looks better :) I have no idea why you added the `type` argument - it's useless. We can call `request.setTimeout` directly.

IMO if the tests were run serially it would be a better approach.
@sindresorhus What do you think?

Reply:
I can undo that change. I was previously injecting the `addTimeout` function rather than the timers object.

Even with serial execution, other modules are affected by globally injecting a fake clock. The tests just hang indefinitely.